## Some header here

Some text here

This notebook is written in **Python** so the default cell type is Python. However, you can use different languages by using the `%LANGUAGE` syntax. Python, Scala, SQL, and R are all supported.

#### We are going to go through some basic statistics we encounter/need when in the process of data exploration

#### 1. Normal Distribution (a.k.a. Gaussian distribution or bell curve)

This is the most common distribution (therefore considered the most important) we encounter when working with various types of data.
But what makes it so commonly used in the natural and social sciences (basically everywhere)? ND is used to represent real-valued random variables whose distributions are not known.

Why it works? How do you know that the distribution of all males per age around  the globe is normal?

Central limit theorem to the rescue - basically it states that under some conditions, the average of many samples (observations) of a random variable with finite mean and variance is itself a random variable whose distribution converges to a normal distribution as the number of samples increases. In other words if you have enough records of a random variable it's distribution is very similar to the normal distribution.

For example the height of people, the blood pressure, errors in measurements, IQ scores - all these very closely resemble the normal distribution.

Cool, isn't it?  🤔

*** pic with IQ distribution

In [4]:
#Insert image here: 

#df = spark.read.format("image").load("https://www.quora.com/What-does-the-distribution-of-IQ-scores-look-like")

#https://www.quora.com/What-does-the-distribution-of-IQ-scores-look-like

the general form of the probability density function of the normal distribution is:

$$ P(x) = \frac{1}{{\sigma \sqrt {2\pi } }}e^{{{ - \left( {x - \mu } \right)^2 } \mathord{\left/ {\vphantom {{ - \left( {x - \mu } \right)^2 } {2\sigma ^2 }}} \right. \kern-\nulldelimiterspace} {2\sigma ^2 }}} $$


If we display all possible values of our random variable (if we think about the IQ example those would be all the values between 60 and 180 roughly) the prob.dnesity finction would give us the relative likelyhood that our ranodom variable (IQ) equals any of the possible values. 

The two parameters here are $$\mu$$ and $$\sigma$$ where mu is the mean (and in this case also the median and the mode) - the probabilistic version of arithmetic mean and is a measure of central tendency of the distribution; the second parameter - sigma - is the standard deviation - a measure of the amount of variation or dispersion of a set of values. A low standard deviation indicates that the values tend to be close to the mean of the set, while a high standard deviation indicates that the values are spread out over a wider range.

If we square the standard devaition we get the variance of a distribution - how far our set is spread out from the average value (the mean value).

I mentioned some statistical features so let's define them as well:
- arithmetic mean (mathematical expectation, average) is the sum of values divided by the number of values
- median is the value separating the higher half from the lower half of a data sample in other words the value exactly in the middle of the set
- mode is the value that appears most often in a set

Knowing that let's take another look at the bell curve and see our statistical features in action:

We see the middle value (our mean) and we see how the deviation tells us how probable it is to be close to this middle value. We also see what percentage of the data is located in different places. The bright blue part of the plot shows that 68.2% of the data is located within one standard deviation away from the middle in each direction. Makes sense - it is most lakely our variable takes the mean value or close to it. As we get further away from the mean we get less and less data. If we look at the IQ chart again we see that the people with IQ higher that 140 are only 2% of the population a.k.a the probability that you are a genious (assuming you are a random person hehe) is only 2%! Sorry to break it to you like that, guys.

We just used another stats feature - percentile - the measure indicating the value below which a given percentage of observations in a group of observations falls.

There are many other distributions with lots of applications like the uniform distribution, the poisson distribution, etc. but we won't get into detail about them right now.

Now that we have revised some basic stats features we might want to visualize them. Boxplots are a great way to do that as they show the basics and gice you a quick overview of the distribution of your data.
There is a boxplot of normally distributed data:


Let's go through the plot - the yellow line - the median- we are already familiar, the Qs are the percentiles: Q1- the 25th percentile, Q3- the 75th, meaning that the body of our boxplot represents 50% of the data;
We have min and max which is (almost) our lowest and highest values. 
The outliers - the extreme low or high values in our data oh and the purple lines are called whiskers.
The IQR (Interquartile range) is simply Q3 - Q1 and determines what will be the min/max values and the outliers.
Finally let's compare the boxplot to the density:


Not all boxplots will look as pretty and symmetrical. Depending on the distribution of the data we might get something like this:




We have all seen these in school maths and most likely hated calculating them which reminds when one of the kids in my class was complaining about the stuff we learn saying "Miss, when are we going to use maths in real life?" and she looked him and replied "Well you certainly won't but some of the smart kids might!". Yeah, my teacher was pretty savage.

Now since we (obviously) are the smart kids let's see how these stats features help us in real life!

We will load one public data set that caught my eye, explore it and use some statistics to gain better understanding of the data.

In [6]:
from pyspark.sql.functions import *
import pandas
import os
import databricks.koalas as ks
import matplotlib.pyplot as plt

In [7]:
# File location and type
file_location = "/FileStore/tables/master.csv"
file_type = "csv"

df = spark.read.csv(
    path = file_location
  , escape = '"'
  , header = True
  , inferSchema = True
  , multiLine = True
  , mode = 'PERMISSIVE'
  , sep = ","
  , quote =  '"'
)


display(df)

country year sex age suicides_no population suicides/100k pop country-year HDI for year gdp_for_year ($) gdp_per_capita ($) generation
 Albania 1987 male 15-24 years 21 312900 6.71 Albania1987 null 2,156,624,900 796 Generation X
 Albania 1987 male 35-54 years 16 308000 5.19 Albania1987 null 2,156,624,900 796 Silent
 Albania 1987 female 15-24 years 14 289700 4.83 Albania1987 null 2,156,624,900 796 Generation X
 Albania 1987 male 75+ years 1 21800 4.59 Albania1987 null 2,156,624,900 796 G.I. Generation
 Albania 1987 male 25-34 years 9 274300 3.28 Albania1987 null 2,156,624,900 796 Boomers
 Albania 1987 female 75+ years 1 35600 2.81 Albania1987 null 2,156,624,900 796 G.I. Generation
 Albania 1987 female 35-54 years 6 278800 2.15 Albania1987 null 2,156,624,900 796 Silent
 Albania 1987 female 25-34 years 4 257200 1.56 Albania1987 null 2,156,624,900 796 Boomers
 Albania 1987 male 55-74 years 1 137500 0.73 Albania1987 null 2,156,624,900 796 G.I. Generation
 Albania 1987 female 5-14 years 0 311000 0.0 Albania1987 null 2,156,624,900 796 Generation X
 Albania 1987 female 55-74 years 0 144600 0.0 Albania1987 null 2,156,624,900 796 G.I. Generation
 Albania 1987 male 5-14 years 0 338200 0.0 Albania1987 null 2,156,624,900 796 Generation X
 Albania 1988 female 75+ years 2 36400 5.49 Albania1988 null 2,126,000,000 769 G.I. Generation
 Albania 1988 male 15-24 years 17 319200 5.33 Albania1988 null 2,126,000,000 769 Generation X
 Albania 1988 male 75+ years 1 22300 4.48 Albania1988 null 2,126,000,000 769 G.I. Generation
 Albania 1988 male 35-54 years 14 314100 4.46 Albania1988 null 2,126,000,000 769 Silent
 Albania 1988 male 55-74 years 4 140200 2.85 Albania1988 null 2,126,000,000 769 G.I. Generation
 Albania 1988 female 15-24 years 8 295600 2.71 Albania1988 null 2,126,000,000 769 Generation X
 Albania 1988 female 55-74 years 3 147500 2.03 Albania1988 null 2,126,000,000 769 G.I. Generation
 Albania 1988 female 25-34 years 5 262400 1.91 Albania1988 null 2,126,000,000 769 Boomers
 Albania 1988 male 25-34 years 5 279900 1.79 Albania1988 null 2,126,000,000 769 Boomers
 Albania 1988 female 35-54 years 4 284500 1.41 Albania1988 null 2,126,000,000 769 Silent
 Albania 1988 female 5-14 years 0 317200 0.0 Albania1988 null 2,126,000,000 769 Generation X
 Albania 1988 male 5-14 years 0 345000 0.0 Albania1988 null 2,126,000,000 769 Generation X
 Albania 1989 male 75+ years 2 22500 8.89 Albania1989 null 2,335,124,988 833 G.I. Generation
 Albania 1989 male 25-34 years 18 283600 6.35 Albania1989 null 2,335,124,988 833 Boomers
 Albania 1989 male 35-54 years 15 318400 4.71 Albania1989 null 2,335,124,988 833 Silent
 Albania 1989 male 55-74 years 6 142100 4.22 Albania1989 null 2,335,124,988 833 G.I. Generation
 Albania 1989 male 15-24 years 12 323500 3.71 Albania1989 null 2,335,124,988 833 Generation X
 Albania 1989 female 35-54 years 7 288600 2.43 Albania1989 null 2,335,124,988 833 Silent
 Albania 1989 female 15-24 years 5 299900 1.67 Albania1989 null 2,335,124,988 833 Generation X
 Albania 1989 female 25-34 years 2 266300 0.75 Albania1989 null 2,335,124,988 833 Boomers
 Albania 1989 female 55-74 years 1 149600 0.67 Albania1989 null 2,335,124,988 833 G.I. Generation
 Albania 1989 female 5-14 years 0 321900 0.0 Albania1989 null 2,335,124,988 833 Generation X
 Albania 1989 female 75+ years 0 37000 0.0 Albania1989 null 2,335,124,988 833 G.I. Generation
 Albania 1989 male 5-14 years 0 349700 0.0 Albania1989 null 2,335,124,988 833 Generation X
 Albania 1992 male 35-54 years 12 343800 3.49 Albania1992 null 709,452,584 251 Boomers
 Albania 1992 male 15-24 years 9 263700 3.41 Albania1992 null 709,452,584 251 Generation X
 Albania 1992 male 55-74 years 5 159500 3.13 Albania1992 null 709,452,584 251 Silent
 Albania 1992 male 25-34 years 7 245500 2.85 Albania1992 null 709,452,584 251 Boomers
 Albania 1992 female 15-24 years 7 292400 2.39 Albania1992 null 709,452,584 251 Generation X
 Albania 1992 female 25-34 years 4 267400 1.5 Albania1992 null 709,452,584 251 Boomers
 Alban

## Basic data exploration

Are there any features that are not relevent? Is our data of good quality? Is there missing data?

In [9]:
from pyspark.sql.functions import sum as _sum, count #So the reason why the build-in function won't work is that's it takes an iterable as an argument where as here the name of the column passed is a string and the built-in function can't be applied on a string

In [10]:
display(df.select('country', 'year').distinct().groupBy('country').agg(count('year')).orderBy('count(year)'))

country count(year) Dominica 1 Macau 1 Mongolia 1 Cabo Verde 1 Bosnia and Herzegovina 2 San Marino 3 Saint Kitts and Nevis 3 Oman 3 United Arab Emirates 6 Nicaragua 6 Turkey 7 Montenegro 10 Maldives 10 Kiribati 11 Fiji 11 Sri Lanka 11 Aruba 14 Cyprus 15 Qatar 15 Philippines 15 Azerbaijan 16 Jamaica 17 Seychelles 18 Serbia 18 South Africa 20 Estonia 21 Belarus 21 Switzerland 21 Latvia 21 Slovenia 21 Bahrain 21 Georgia 22 Croatia 22 Slovakia 22 Uzbekistan 22 Lithuania 22 Albania 22 Denmark 22 Bahamas 23 Poland 24 Cuba 24 El Salvador 24 Kuwait 25 Guyana 25 Saint Vincent and Grenadines 25 Panama 25 Armenia 25 Barbados 25 Grenada 26 Hungary 26 Kyrgyzstan 26 Germany 26 Kazakhstan 26 Paraguay 27 Trinidad and Tobago 27 Czech Republic 27 Antigua and Barbuda 27 Russian Federation 27 Portugal 27 Uruguay 28 Ukraine 28 Suriname 28 Romania 28 Thailand 28 Saint Lucia 28 Belize 28 Finland 29 Turkmenistan 29 Canada 29 New Zealand 29 Ireland 30 France 30 Bulgaria 30 Sweden 30 Norway 30 Costa Rica 30 Australia 30 Guatemala 30 United States 31 Singapore 31 Chile 31 Italy 31 Spain 31 Republic of Korea 31 Argentina 31 Malta 31 Mexico 31 Puerto Rico 31 Luxembourg 31 Greece 31 United Kingdom 31 Ecuador 31 Belgium 31 Brazil 31 Japan 31 Israel 31 Colombia 31 Iceland 32 Mauritius 32 Austria 32 Netherlands 32

In [11]:
display(df.select('year','country').distinct().groupBy('year').agg(count('country')).orderBy('year'))

year count(country) 1985 48 1986 48 1987 54 1988 49 1989 52 1990 64 1991 64 1992 65 1993 65 1994 68 1995 78 1996 77 1997 77 1998 79 1999 83 2000 86 2001 88 2002 86 2003 86 2004 84 2005 84 2006 85 2007 86 2008 85 2009 89 2010 88 2011 86 2012 81 2013 80 2014 78 2015 62 2016 16

Based on the data we have we are going to get rid of the years with too few records

In [13]:
df_new = df.filter(col('year').between('1995','2014'))

In [14]:
df_new.select('year').describe().show()

+-------+-----------------+
summary| year|
+-------+-----------------+
 count| 19992|
 mean|2004.578031212485|
 stddev|5.672697271834574|
 min| 1995|
 max| 2014|
+-------+-----------------+

Let's explore the gobal averages by country and year

In [16]:
display(df_new.select('year','suicides/100k pop').groupBy('year').agg(avg('suicides/100k pop')).orderBy('year'))

year avg(suicides/100k pop) 1995 15.662670940170946 1996 15.305422077922088 1997 14.954361471861489 1998 14.926919831223637 1999 14.532038152610449 2000 13.94132751937986 2001 13.519138257575767 2002 13.786550387596913 2003 13.20501937984495 2004 12.481944444444451 2005 12.068442460317463 2006 11.927460784313718 2007 12.025339147286827 2008 11.907686274509805 2009 11.400786516853934 2010 11.215899621212124 2011 11.015348837209324 2012 11.42171810699588 2013 11.107958333333325 2014 11.011463675213678

We can clearly see that globaly there is a tendency of lowering the rates per 100k pop!

In [18]:
display(df_new.select('country','suicides/100k pop').groupBy('country').agg(avg('suicides/100k pop')).orderBy('country'))

country avg(suicides/100k pop) Albania 3.8766145833333336 Antigua and Barbuda 0.68375 Argentina 10.112999999999996 Armenia 2.984907407407407 Aruba 9.503095238095236 Australia 12.289210526315786 Austria 21.026499999999977 Azerbaijan 1.674696969696969 Bahamas 1.2185087719298244 Bahrain 1.8362962962962963 Barbados 1.1808333333333336 Belarus 32.16765624999999 Belgium 19.80341666666667 Belize 6.988500000000002 Bosnia and Herzegovina 4.617083333333333 Brazil 6.0072499999999955 Bulgaria 17.257 Cabo Verde 11.153333333333334 Canada 11.553201754385958 Chile 11.755249999999997 Colombia 6.029833333333333 Costa Rica 7.19983333333333 Croatia 23.081666666666663 Cuba 19.96537500000001 Cyprus 3.137948717948717 Czech Republic 17.166666666666675 Denmark 13.945874999999992 Ecuador 6.411416666666663 El Salvador 10.149458333333342 Estonia 27.785 Fiji 5.105000000000001 Finland 20.421916666666657 France 19.297333333333327 Georgia 4.002156862745097 Germany 14.460249999999986 Greece 3.766833333333337 Grenada 2.1054687500000004 Guatemala 3.1487916666666673 Guyana 25.50027777777777 Hungary 30.98795833333332 Iceland 12.497666666666671 Ireland 11.272583333333335 Israel 7.974958333333336 Italy 7.785749999999998 Jamaica 0.6885833333333333 Japan 21.573041666666672 Kazakhstan 31.30841666666665 Kiribati 3.1020238095238093 Kuwait 1.221 Kyrgyzstan 13.044791666666685 Latvia 29.674500000000016 Lithuania 41.006999999999984 Luxembourg 15.199666666666669 Maldives 1.3673333333333335 Malta 5.793624999999998 Mauritius 10.724249999999994 Mexico 5.051125 Montenegro 9.950249999999999 Netherlands 9.792333333333334 New Zealand 13.866622807017546 Nicaragua 6.715833333333333 Norway 11.935041666666674 Oman 0.7361111111111112 Panama 6.160745614035086 Paraguay 4.631874999999999 Philippines 2.5900641025641034 Poland 15.321666666666669 Portugal 10.736715686274508 Puerto Rico 8.73183333333333 Qatar 1.813333333333332 Republic of Korea 31.74229166666667 Romania 12.523208333333336 Russian Federation 34.77766666666667 Saint Lucia 7.0062719298245595 Saint Vincent and Grenadines 6.640416666666667 San Marino 4.045 Serbia 22.45823529411764 Seychelles 7.704107142857144 Singapore 14.071000000000002 Slovakia 11.945394736842106 Slovenia 28.115874999999992 South Africa 0.9582894736842105 Spain 9.107291666666665 Sri Lanka 33.059629629629626 Suriname 21.826833333333333 Sweden 13.423458333333325 Switzerland 19.289208333333345 Thailand 7.220087719298246 Trinidad and Tobago 14.159411764705888 Turkey 2.3708333333333336 Turkmenistan 7.715458333333334 Ukraine 26.913859649122813 United Arab Emirates 1.3179166666666662 United Kingdom 6.940291666666664 United States 13.01204166666667 Uruguay 20.942192982456127 Uzbekistan 7.601617647058826

In [19]:
display(df_new.select('age','suicides/100k pop').groupBy('age').agg(avg('suicides/100k pop')).orderBy('age'))

age avg(suicides/100k pop) 15-24 years 8.99929171668668 25-34 years 12.33504501800722 35-54 years 15.32968787515006 5-14 years 0.6364825930372141 55-74 years 16.260357142857124 75+ years 23.47780012004808

Globally half the suicides are commited by people above 55 years old

In [21]:
dfiso = ks.read_excel("/dbfs/FileStore/tables/ISO_Alpha3_Codes.xlsx") ## if you don't have dbfs/.. in the path read_excel does not find the file!!! Fun stuff

dfISO = dfiso.to_spark()

# try to display some map here

In [22]:
display(dfISO)

UN Trade Statistics Abbreviation Country ABW Aruba ABW Aruba AFG Afghanistan AFG Afghanistan AGO Angola AGO Angola AIA Anguilla AIA Anguilla ALA Åland Islands ALA Åland Islands ALB Albania ALB Albania AND Andorra AND Andorra ANT Netherlands Antilles ANT Netherlands Antilles ARE United Arab Emirates ARE United Arab Emirates ARG Argentina ARG Argentina ARM Armenia ARM Armenia ASM American Samoa ASM American Samoa ATA Antarctica ATA Antarctica ATF French Southern Territories ATF French Southern Territories ATG Antigua and Barbuda ATG Antigua and Barbuda AUS Australia AUS Australia AUT Austria AUT Austria AZE Azerbaijan AZE Azerbaijan BDI Burundi BDI Burundi BEL Belgium BEL Belgium BEN Benin BEN Benin BFA Burkina Faso BFA Burkina Faso BGD Bangladesh BGD Bangladesh BGR Bulgaria BGR Bulgaria BHR Bahrain BHR Bahrain BHS Bahamas BHS Bahamas BIH Bosnia and Herzegovina BIH Bosnia and Herzegovina BLM Saint Barthélemy BLM Saint Barthélemy BLR Belarus BLR Belarus BLZ Belize BLZ Belize BMU Bermuda BMU Bermuda BOL Bolivia, Plurinational State of BOL Bolivia, Plurinational State of BRA Brazil BRA Brazil BRB Barbados BRB Barbados BRN Brunei Darussalam BRN Brunei Darussalam BTN Bhutan BTN Bhutan BVT Bouvet Island BVT Bouvet Island BWA Botswana BWA Botswana CAF Central African Republic CAF Central African Republic CAN Canada CAN Canada CCK Cocos (Keeling) Islands CCK Cocos (Keeling) Islands CHE Switzerland CHE Switzerland CHL Chile CHL Chile CHN China CHN China CIV Côte d'Ivoire CIV Côte d'Ivoire CMR Cameroon CMR Cameroon COD Congo, the Democratic Republic of the COD Congo, the Democratic Republic of the COG Congo COG Congo COK Cook Islands COK Cook Islands COL Colombia COL Colombia COM Comoros COM Comoros CPV Cape Verde CPV Cape Verde CRI Costa Rica CRI Costa Rica CUB Cuba CUB Cuba CXR Christmas Island CXR Christmas Island CYM Cayman Islands CYM Cayman Islands CYP Cyprus CYP Cyprus CZE Czech Republic CZE Czech Republic DEU Germany DEU Germany DJI Djibouti DJI Djibouti DMA Dominica DMA Dominica DNK Denmark DNK Denmark DOM Dominican Republic DOM Dominican Republic DZA Algeria DZA Algeria ECU Ecuador ECU Ecuador EGY Egypt EGY Egypt ERI Eritrea ERI Eritrea ESH Western Sahara ESH Western Sahara ESP Spain ESP Spain EST Estonia EST Estonia ETH Ethiopia ETH Ethiopia FIN Finland FIN Finland FJI Fiji FJI Fiji FLK Falkland Islands (Malvinas) FLK Falkland Islands (Malvinas) FRA France FRA France FRO Faroe Islands FRO Faroe Islands FSM Micronesia, Federated States of FSM Micronesia, Federated States of GAB Gabon GAB Gabon GBR United Kingdom GBR United Kingdom GEO Georgia GEO Georgia GGY Guernsey GGY Guernsey GHA Ghana GHA Ghana GIB Gibraltar GIB Gibraltar GIN Guinea GIN Guinea GLP Guadeloupe GLP Guadeloupe GMB Gambia GMB Gambia GNB Guinea-Bissau GNB Guinea-Bissau GNQ Equatorial Guinea GNQ Equatorial Guinea GRC Greece GRC Greece GRD Grenada GRD Grenada GRL Greenland GRL Greenland GTM Guatemala GTM Guatemala GUF French Guiana GUF French Guiana GUM Guam GUM Guam GUY Guyana GUY Guyana HKG Hong Kong HKG Hong Kong HMD Heard Island and McDonald Islands HMD Heard Island and McDonald Islands HND Honduras HND Honduras HRV Croatia HRV Croatia HTI Haiti HTI Haiti HUN Hungary HUN Hungary IDN Indonesia IDN Indonesia IMN Isle of Man IMN Isle of Man IND India IND India IOT British Indian Ocean Territory IOT British Indian Ocean Territory IRL Ireland IRL Ireland IRN Iran, Islamic Republic of IRN Iran, Islamic Republic of IRQ Iraq IRQ Iraq ISL Iceland ISL Iceland ISR Israel ISR Israel ITA Italy ITA Italy JAM Jamaica JAM Jamaica JEY Jersey JEY Jersey JOR Jordan JOR Jordan JPN Japan JPN Japan KAZ Kazakhstan KAZ Kazakhstan KEN Kenya KEN Kenya KGZ Kyrgyzstan KGZ Kyrgyzstan KHM Cambodia KHM Cambodia KIR Kiribati KIR Kiribati KNA Saint Kitts and Nevis KNA Saint Kitts and Nevis KOR Korea, Republic of KOR Korea, Republic of KWT Kuwait KWT Kuwait LAO Lao People's Democratic Republic LAO Lao People's Democratic Republic LBN Lebanon LBN Lebanon LBR Liberia LBR Liberia LBY Libyan Arab J

In [23]:
df_to_map = df_new.alias('t1').join(dfISO.alias('t2'), lower(col('t1.country')) == lower(col('t2.Country')), 'left')

In [24]:
display(df_to_map.select('t1.country').distinct().filter(col('Abbreviation').isNull() == True))

country Cabo Verde Republic of Korea Saint Vincent and Grenadines

In [25]:
display(df_new.select('country').distinct().filter(col('Country').like('%Korea%')))

country Republic of Korea

The issue we encounter is that in the dataset we have only one Republic of Korea, now since we only need this to be able to visualise the countries on the map let's say it is PRK

In [27]:
display(dfISO.filter(col('Country').like('%Grenadines%')))

UN Trade Statistics Abbreviation Country VCT Saint Vincent and the Grenadines VCT Saint Vincent and the Grenadines

In [28]:
df_to_map =(df_to_map.withColumn('Abbreviation', when(col('t1.country') == 'Republic of Korea', 'PRK')
                                                .when(col('t1.country') == 'Cabo Verde', 'CPV')
                                                .when(col('t1.country') == 'Saint Vincent and Grenadines', 'VCT')
                                                .otherwise(col('Abbreviation'))))

In [29]:
display(df_to_map.select('t1.country').distinct().filter(col('Abbreviation').isNull() == True))

country

#### Cool no more missing abbrevaitions, no we are ready to use map

In [31]:
display(df_to_map.select('Abbreviation','suicides/100k pop').groupBy('Abbreviation').agg(avg('suicides/100k pop')).orderBy('Abbreviation'))

Abbreviation avg(suicides/100k pop) ABW 9.503095238095236 ALB 3.8766145833333336 ARE 1.3179166666666662 ARG 10.112999999999996 ARM 2.984907407407407 ATG 0.68375 AUS 12.289210526315786 AUT 21.026499999999977 AZE 1.674696969696969 BEL 19.80341666666667 BGR 17.257 BHR 1.8362962962962963 BHS 1.2185087719298244 BIH 4.617083333333333 BLR 32.16765624999999 BLZ 6.988500000000002 BRA 6.0072499999999955 BRB 1.1808333333333336 CAN 11.553201754385958 CHE 19.289208333333345 CHL 11.755249999999997 COL 6.029833333333333 CPV 11.153333333333334 CRI 7.19983333333333 CUB 19.96537500000001 CYP 3.137948717948717 CZE 17.166666666666675 DEU 14.460249999999986 DNK 13.945874999999992 ECU 6.411416666666663 ESP 9.107291666666665 EST 27.785 FIN 20.421916666666657 FJI 5.105000000000001 FRA 19.297333333333327 GBR 6.940291666666664 GEO 4.002156862745097 GRC 3.766833333333337 GRD 2.1054687500000004 GTM 3.1487916666666673 GUY 25.50027777777777 HRV 23.081666666666663 HUN 30.98795833333332 IRL 11.272583333333335 ISL 12.497666666666671 ISR 7.974958333333336 ITA 7.785749999999998 JAM 0.6885833333333333 JPN 21.573041666666672 KAZ 31.30841666666665 KGZ 13.044791666666685 KIR 3.1020238095238093 KWT 1.221 LCA 7.0062719298245595 LKA 33.059629629629626 LTU 41.006999999999984 LUX 15.199666666666669 LVA 29.674500000000016 MDV 1.3673333333333335 MEX 5.051125 MLT 5.793624999999998 MNE 9.950249999999999 MUS 10.724249999999994 NIC 6.715833333333333 NLD 9.792333333333334 NOR 11.935041666666674 NZL 13.866622807017546 OMN 0.7361111111111112 PAN 6.160745614035086 PHL 2.5900641025641034 POL 15.321666666666669 PRI 8.73183333333333 PRK 31.74229166666667 PRT 10.736715686274508 PRY 4.631874999999999 QAT 1.813333333333332 ROU 12.523208333333336 RUS 34.77766666666667 SGP 14.071000000000002 SLV 10.149458333333342 SMR 4.045 SRB 22.45823529411764 SUR 21.826833333333333 SVK 11.945394736842106 SVN 28.115874999999992 SWE 13.423458333333325 SYC 7.704107142857144 THA 7.220087719298246 TKM 7.715458333333334 TTO 14.159411764705888 TUR 2.3708333333333336 UKR 26.913859649122813 URY 20.942192982456127 USA 13.01204166666667 UZB 7.601617647058826 VCT 6.640416666666667 ZAF 0.9582894736842105

In [32]:
kdf = ks.DataFrame(df_to_map.select('t1.*', 'Abbreviation'))

In [33]:
kdf.describe() ## compare to the more vanilla describe() method of pyspark

year 
 suicides_no 
 population 
 suicides/100k pop 
 HDI for year 
 gdp_per_capita ($) 
 
 
 
 
 count 
 19992.000000 
 19992.000000 
 1.999200e+04 
 19992.000000 
 7320.000000 
 19992.000000 
 
 
 mean 
 2004.578031 
 241.489096 
 1.820575e+06 
 12.839777 
 0.786282 
 18968.867347 
 
 
 std 
 5.672697 
 919.405021 
 3.936059e+06 
 19.231296 
 0.090211 
 20324.160757 
 
 
 min 
 1995.000000 
 0.000000 
 2.780000e+02 
 0.000000 
 0.513000 
 291.000000 
 
 
 25% 
 2000.000000 
 3.000000 
 9.729900e+04 
 0.920000 
 0.721000 
 4095.000000 
 
 
 50% 
 2005.000000 
 24.000000 
 4.295000e+05 
 5.770000 
 0.786000 
 11062.000000 
 
 
 75% 
 2009.000000 
 125.000000 
 1.455000e+06 
 16.430000 
 0.866000 
 28722.000000 
 
 
 max 
 2014.000000 
 21706.000000 
 4.380521e+07 
 224.970000 
 0.944000 
 126352.000000

In [34]:
kdf.isnull().sum()

Out[24]: country 0
year 0
sex 0
age 0
suicides_no 0
population 0
suicides/100k pop 0
country-year 0
HDI for year 12672
 gdp_for_year ($) 0
gdp_per_capita ($) 0
generation\r 0
Abbreviation 0
dtype: int64

In [35]:
kdf.fillna(0)
kdf.head()

country 
 year 
 sex 
 age 
 suicides_no 
 population 
 suicides/100k pop 
 country-year 
 HDI for year 
 gdp_for_year ($) 
 gdp_per_capita ($) 
 generation 
 Abbreviation 
 
 
 
 
 0 
 Aruba 
 1995 
 male 
 75+ years 
 2 
 889 
 224.97 
 Aruba1995 
 NaN 
 1,320,670,391 
 17949 
 G.I. Generation\r 
 ABW 
 
 
 1 
 Aruba 
 1995 
 male 
 25-34 years 
 4 
 7520 
 53.19 
 Aruba1995 
 NaN 
 1,320,670,391 
 17949 
 Generation X\r 
 ABW 
 
 
 2 
 Aruba 
 1995 
 male 
 55-74 years 
 1 
 4558 
 21.94 
 Aruba1995 
 NaN 
 1,320,670,391 
 17949 
 Silent\r 
 ABW 
 
 
 3 
 Aruba 
 1995 
 male 
 35-54 years 
 1 
 11895 
 8.41 
 Aruba1995 
 NaN 
 1,320,670,391 
 17949 
 Boomers\r 
 ABW 
 
 
 4 
 Aruba 
 1995 
 female 
 15-24 years 
 0 
 5138 
 0.00 
 Aruba1995 
 NaN 
 1,320,670,391 
 17949 
 Generation X\r 
 ABW

In [36]:
import seaborn as sns
from scipy.stats import norm

f, (ax1, ax2, ax3) = plt.subplots(1,3, figsize=(20, 6))

dist = kdf['population'].loc[kdf['suicides/100k pop'] == 1].values
sns.distplot(dist,ax=ax1, fit=norm, color='#FB8861')
ax1.set_title('title', fontsize=14)

# v12_fraud_dist = new_df['V12'].loc[new_df['Class'] == 1].values
# sns.distplot(v12_fraud_dist,ax=ax2, fit=norm, color='#56F9BB')
# ax2.set_title('V12 Distribution \n (Fraud Transactions)', fontsize=14)

plt.show()

--------------------------------------------------------------------------- 
 PandasNotImplementedError Traceback (most recent call last)
 <command-2917089042630018> in <module> 
 4 f , ( ax1 , ax2 , ax3 ) = plt . subplots ( 1 , 3 , figsize = ( 20 , 6 ) ) 
 5 
 ----> 6 dist = kdf [ 'population' ] . loc [ kdf [ 'suicides/100k pop' ] == 1 ] . values
 7 sns . distplot ( dist , ax = ax1 , fit = norm , color = '#FB8861' ) 
 8 ax1 . set_title ( 'title' , fontsize = 14 ) 

 /databricks/python/lib/python3.7/site-packages/databricks/koalas/usage_logging/__init__.py in wrapper (*args, **kwargs) 
 138 start = time . perf_counter ( ) 
 139 try : 
 --> 140 res = func ( * args , ** kwargs ) 
 141 logger.log_success(
 142 class_name, function_name, time.perf_counter() - start, signature)

 /databricks/python/lib/python3.7/site-packages/databricks/koalas/series.py in __getattr__ (self, item) 
 4271 property_or_func = getattr ( _MissingPandasLikeSeries , item ) 
 4272 if isinstance ( property_or_func , property ) : 
 -> 4273 return property_or_func . fget ( self ) # type: ignore 
 4274 else : 
 4275 return partial ( property_or_func , self ) 

 /databricks/python/lib/python3.7/site-packages/databricks/koalas/usage_logging/__init__.py in wrapper (self) 
 208 def wrapper ( self ) : 
 209 try : 
 --> 210 return prop . fget ( self ) 
 211 finally : 
 212 logger . log_missing ( class_name , property_name , is_deprecated ) 

 /databricks/python/lib/python3.7/site-packages/databricks/koalas/missing/__init__.py in unsupported_property (self) 
 36 def unsupported_property ( self ) : 
 37 raise PandasNotImplementedError(class_name=class_name, property_name=property_name,
 ---> 38 reason=reason)
 39 
 40 @ property

 PandasNotImplementedError : The property `pd.Series.values()` is not implemented. If you want to collect your data as an NumPy array, use 'to_numpy()' instead.

## Basic statistical features

When you are exploring data it is common to look for patterns that can help you derive some insights : what is the most important feature of my data that impact my sales, what features to leave if i want to predict my sales values? And so on...

Here is when the basic statistics come in hand. 

Let's go through some commonly used stat methods and apply them to our dataset

###### 1. Correlations

We can easily display correlations between variables (features of our dataset) using the correlation matrix which is nothing different that a table showing the correlation coefficients between the variables.

The correlation coefficient itself is represented by the formula:

#### Mean - arithmetic mean, also called the mathematical expectation or average, is the central value of a discrete set of numbers: specifically, the sum of the values divided by the number of values

Let's see the mean s.rate per country, rate and age

In [42]:
df.describe().show()

# stdev tell how measurements for a group are spread out from the average (mean), or expected value. A low standard deviation means that most of the numbers are close to the average. A high standard deviation means that the numbers are more spread out.it is the square root of the Variance.

# The variance is the average of the squared differences from the Mean.

In [43]:
import matplotlib.pyplot as plt

f, axes = plt.subplots(ncols=5, figsize=(20,4))
colors = ["#0101DF", "#DF0101"]

# Negative Correlations with our Class (The lower our feature value the more likely it will be a fraud transaction)
sns.boxplot(x="sex", y="suicides/100k pop", data=kdf, palette=colors, ax=axes[0])
axes[0].set_title('Some title')

In [44]:
ax = kdf[''].plot.box()

In [45]:
df = spark.read.format("image").load("")

#### What are boxplots?

- way of displaying the distribution of data

#### Characteristics of a boxplot

- median (Q2/50th Percentile): the middle value of the dataset.
- first quartile (Q1/25th Percentile): the middle number between the smallest number (not the “minimum”) and the median of the dataset
- third quartile (Q3/75th Percentile): the middle value between the median and the highest value (not the “maximum”) of the dataset.
- interquartile range (IQR): 25th to the 75th percentile.
- whiskers 
- outliers
- “maximum”: Q3 + 1.5*IQR
- “minimum”: Q1 -1.5*IQR

#### Boxplot on a Normal Distribution - for better understanding

insert some picture once you figure out how

What's the big deal with normal distribution?

A PDF(probability density function) is used to specify the probability of the random variable falling within a particular range of values, as opposed to taking on any one value